tiny_dnn/util/parallel_for.h

```diff
@@ -96,61 +91,113 @@ void parallel_for(int begin, int end, const Func &f, int /*grainsize*/) {
 
 #else
 
-template <typename Func>
-void parallel_for(int start, int end, const Func &f, int /*grainsize*/) {
-  static int nthreads = std::thread::hardware_concurrency();
-
-#ifdef DNN_USE_THREADPOOL
-  static ThreadPool pool(nthreads);
-
-  std::vector<std::future<void> > futures;
-  int total_size = end - start;
-  if (total_size <= nthreads) {
-    for (int i = 0; i < total_size; i++) {
-      int block_start = start + i;
-      int block_end   = start + i + 1;
-      futures.emplace_back(pool.enqueue([block_start, block_end, &f] {
-        f(blocked_range(block_start, block_end));
-      }));
+struct thread_pool {
+  thread_pool(size_t nthreads = std::thread::hardware_concurrency())
```

Comment:
What about moving `thread_pool` and `singleton` into `tiny_dnn::detail` namespace?